New feature: Local distribution capability

JackLau1222 opened the pull request:
I'm excited to introduce a new feature that enables local distribution capability for BMF, allowing users to easily 
parallelize computations on a single node.

In this commit, I've implemented the necessary code changes to support multi-threading for single node. To enable this 
feature, users can simply add an option parameter `"thread":<number>` when setting up these nodes, and the framework will create multiple distributed nodes to process data in parallel.

**Key Changes:**

* update enhance_demo.py
* add internal module: split_module and assemble_module
* add copy constructors in ModuleConfig, StreamConfig, NodeMetaInfo, NodeConfig
* add methods and variables in NodeConfig
* add create_split_node, create_assemble_node, process_multi_thread functions in Optimizer
* add SplitOutputStreamManager
* update node.cpp
* add new parameter scheduler in the Decode and Encode method in builder
* update connector.cpp
* add new test cases distributed_node

sfeiwong commented on bmf/demo/video_enhance/enhance_demo.py:
```diff
@@ -20,6 +20,7 @@ def run():
         option={
             "fp32": True,
             "output_scale": 2,
+            "thread": 3
```
how about using "distribution" which may be more matched with what it is?

JackLau1222 replied:
good idea! the "distribution" parameter is easy to use, but would "distributed_nums" be more appropriate?

sfeiwong replied:
yes, or "dist_nums" should be shorter?

JackLau1222 replied:
completed!

JackLau1222 commented:
I changed some code to make it more concise

**Key changes:**
- renamed paramater "thread" to "dist_nums"
- optimized process_distributed_node()
- modified some comments
- deleted the subclass of OutputStreamManager
- optimized the distributed node test case

HuHeng commented on bmf/engine/c_engine/src/split_module.cpp:
```diff
@@ -0,0 +1,91 @@
+            if (in_eof_[input_queue.first] == true)
+                continue;
+            // fill splitted pkt into multi output stream
+            task.fill_output_packet(stream_index_, pkt);
```
这里如果是eof packet， 这个eof packet只会往其中一个output_stream送出？其它的output_streams得不到eof packet？

JackLau1222 replied:
确实是这样的，之前把拆分实现在ISM中会缓冲到一个临时队列，检测到eof_packet后会生成`tem_queue_.size()-1`个eof_packet进入临时队列。
迁移到split_module后没有使用缓冲队列，虽然目前这样实现测试中没有遇到错误，但确实是个隐性的问题，我处理一下

JackLau1222 replied:
找到了为什么目前这种实现没有遇到严重的error
```bash
[2024-10-29 20:51:29.373] [info] node id:4 scheduling failed, all downstream node closed: copy_module
[2024-10-29 20:51:29.373] [info] node id:5 close node
[2024-10-29 20:51:29.373] [info] node 5 close report, closed count: 7
[2024-10-29 20:51:29.373] [info] node id:5 scheduling failed, all downstream node closed: copy_module
[2024-10-29 20:51:29.373] [info] node id:6 scheduling failed, all downstream node closed: assemble_module
```
当下流节点接受到eof关闭时，上流节点会自动关闭

JackLau1222 replied:
已经重新在split_module中加入缓冲队列，解决了上述问题

JackLau1222 replied:
考虑到split_module此处只是需要多推送几个eof，没必要加入缓冲队列，于是做了以下修改，更加精简一些
```
task.fill_output_packet(stream_index_, pkt);
if (pkt.timestamp() == BMF_EOF) {
    // fill eof packet for extra distributed node
    for (size_t i = 1; i < task.get_outputs().size(); i++) {
        task.fill_output_packet(i, Packet::generate_eof_packet());
    }
    in_eof_ = true;
}
```

HuHeng commented on bmf/engine/c_engine/include/assemble_module.h:
```diff
@@ -0,0 +1,46 @@
+
+    int queue_index_;
+
+    std::map<int, std::shared_ptr<bmf_engine::SafeQueue<Packet>>> queue_map_;
```
这里应该没必要使用SafeQueue数据结构，模块内部操作queue_map_, 且不涉及多线程访问，使用普通的Queue即可？

JackLau1222 replied:
在这里使用SafeQueue确实显得冗余了，当初仅仅是为了方便pop操作
```cpp
if (queue->second->pop(packet)) {
    task.fill_output_packet(0, packet);
    if (packet.timestamp() == BMF_EOF) {
        in_eof_[queue_index_] = true;
    }
    ...
}
```
目前已推送最新代码
